module/validation/receipt_validator.go

```diff
@@ -0,0 +1,188 @@
+// at the given block and returns the corresponding node's full identity.
+// Error returns:
+//   * sentinel engine.InvalidInputError is nodeID is NOT an authorized member of the network
+//   * generic error indicating a fatal internal problem
+func (v *receiptValidator) identityForNode(blockID flow.Identifier, nodeID flow.Identifier) (*flow.Identity, error) {
+	// get the identity of the origin node
+	identity, err := v.state.AtBlockID(blockID).Identity(nodeID)
+	if err != nil {
+		if protocol.IsIdentityNotFound(err) {
+			return nil, engine.NewInvalidInputErrorf("unknown node identity: %w", err)
+		}
+		// unexpected exception
+		return nil, fmt.Errorf("failed to retrieve node identity: %w", err)
+	}
+
+	return identity, nil
+}
+
+func (v *receiptValidator) verifySignature(receipt *flow.ExecutionReceipt, nodeIdentity *flow.Identity) error {
+	id := receipt.ID()
+	valid, err := v.verifier.Verify(id[:], receipt.ExecutorSignature, nodeIdentity.StakingPubKey)
+	if err != nil {
+		return fmt.Errorf("failed to verify signature: %w", err)
+	}
+
+	if !valid {
+		return engine.NewInvalidInputErrorf("Invalid signature for (%x)", nodeIdentity.NodeID)
+	}
+
+	return nil
+}
+
+func (v *receiptValidator) verifyChunksFormat(result *flow.ExecutionResult) error {
+	for index, chunk := range result.Chunks.Items() {
+		if uint(index) != chunk.CollectionIndex {
+			return engine.NewInvalidInputErrorf("invalid CollectionIndex, expected %d got %d", index, chunk.CollectionIndex)
+		}
+
+		if chunk.BlockID != result.BlockID {
+			return engine.NewInvalidInputErrorf("invalid blockID, expected %v got %v", result.BlockID, chunk.BlockID)
+		}
+	}
+
+	// we create one chunk per collection, plus the
+	// system chunk. so we can check if the chunk number matches with the
+	// number of guarantees plus one; this will ensure the execution receipt
+	// cannot lie about having less chunks and having the remaining ones
+	// approved
+	requiredChunks := 1 // system chunk: must exist for block's ExecutionResult, even if block payload itself is empty
+
+	index, err := v.index.ByBlockID(result.BlockID)
+	if err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return err
+		}
+		// reaching this line means the block is empty, i.e. it has no payload => we expect only the system chunk
+	} else {
+		requiredChunks += len(index.CollectionIDs)
+	}
+
+	if result.Chunks.Len() != requiredChunks {
+		return engine.NewInvalidInputErrorf("invalid number of chunks, expected %d got %d",
+			requiredChunks, result.Chunks.Len())
+	}
+
+	return nil
+}
+
+// subgraphCheck enforces that result forms a valid sub-graph:
+// Let R1 be a result that references block A, and R2 be R1's parent result.
+// The execution results form a valid subgraph if and only if R2 references
+// A's parent.
+func (v *receiptValidator) subgraphCheck(result *flow.ExecutionResult) error {
+	prevResult, err := v.results.ByID(result.PreviousResultID)
```

Comment:
What guarantees we have the results for `all` blocks in the storage?

The matching will store the result into the database when a receipt has been received and passed certain checks. However, this doesn't guarantee a consensus node have received the results for all the blocks, because each execution node only select a few consensus nodes to broadcast the receipts. 

It also doesn't seem to be the protocol state's `Extend` function is storing the result, did I miss anything?

Reply:
Sorry, this is a bit of a subtle details. We should probably document it better:
* The `mutator`s store the block https://github.com/onflow/flow-go/blob/02803f99ffa185e22ddecaececa2075f8cf11345/state/protocol/badger/mutator.go#L650
* the storage implementation of `Blocks` stores the payload as a separate entity https://github.com/onflow/flow-go/blob/02803f99ffa185e22ddecaececa2075f8cf11345/storage/badger/blocks.go#L40
* where the payload is internally stored component by component (i.e. it also stores the receipts individually): https://github.com/onflow/flow-go/blob/02803f99ffa185e22ddecaececa2075f8cf11345/storage/badger/payloads.go#L52-L58
* and finally the receipts storage layer breaks up the receipt in the meta information and the result, where both are stored separately https://github.com/onflow/flow-go/blob/02803f99ffa185e22ddecaececa2075f8cf11345/storage/badger/receipts.go#L32

Furthermore, the matching engine also persists the `result` of any receipt that passes validation: https://github.com/onflow/flow-go/blob/32081ff9f91d24fd5c10ab57bdf447a70c7dd4ef/engine/consensus/matching/engine.go#L274-L276 

#### TLDR
* The storage layer has the `Result`s  incorporated in any known block or any received Execution Receipt (provided they pass validation)
* You are right that we discard Receipts for which we don't have the block or the parent result they reference. However, I don't think this is a problem for exactly the same reason explained in the [comment above](https://github.com/onflow/flow-go/pull/247#discussion_r545573100): we will eventually re-fetch them or other consensus nodes will include them in blocks. 🤞

Reply:
Let's say there are blocks `A <- B <- C (ER_A) <- D (ER_B) <- E`, `C` contains result for `A`.

So the happy case is: if we receive ER_C, whose parent result ID is the ID of  `ER_B`, then we will accept it

And if we receives `ER_D`, whose parent result ID is ID of `ER_C`, even if C is known, we will still drop this receipt, because we haven't receive ER_C yet. And we will rely on the `requestPending` to fetch it.